refactor: support sourcing images from either file path or in-memory data frame

jimthompson5802 opened the pull request:
# Code Pull Requests

PR addresses Issue #484 and Issue #268.

Adds option to support sourcing images from file path or  from in-memory as numpy arrays.

w4nderlust commented on ludwig/features/image_feature.py:
```diff
@@ -294,16 +313,19 @@ def add_feature_data(
-                                  for file_path in input_df[feature[NAME]]]
+                all_img_stores = [get_abs_path(src_path, img_store)
+                                  if isinstance(img_store, str) else img_store
+                                  for img_store in input_df[feature[NAME]]]
```
This should most likely be `in input_df[feature[COLUMN]]` instead. It could have been a leftover error from when we introduced the COLUMN concept

jimthompson5802 replied:
changed to `input_df[feature[COLUMN]]`

w4nderlust commented on ludwig/features/image_feature.py:
```diff
@@ -313,13 +335,16 @@ def add_feature_data(
-                    lambda file_path: read_image_and_resize(get_abs_path(src_path, file_path))
+                    lambda img_store: read_image_and_resize(
+                        get_abs_path(src_path, img_store))
+                    if isinstance(img_store, str) else img_store
```
shouldn't this be `else read_image_and_resize(img_store)`?
Also this is a pretty long lambda, maybe we can define a litle local helper function for clarity ;)

jimthompson5802 replied:
Good catch...you're correct I missed this one.  
created helper function `_get_processed_image()`

w4nderlust commented on ludwig/features/image_feature.py:
```diff
@@ -313,13 +335,16 @@ def add_feature_data(
-                              for file_path in input_df[feature[NAME]]]
+            all_img_stores = [get_abs_path(src_path, img_store)
+                              if isinstance(img_store, str) else img_store
+                              for img_store in input_df[feature[NAME]]]
```
Also here, should likely be COLUMN not NAME

jimthompson5802 replied:
Done

jimthompson5802 commented:
Ready for review.  summary of changes:
* Major changes in `image_feature.py` to support both file paths and ndarray as image features in a pandas data frame.
* Modified api signature for couple core apis to support passing `skip_save_processed_input` flag to the image_feature add_feature_data() to handle different setup situations.
* added unit test to test for file path and ndarray sourcing of images
* added 3 helper functions to utils.data_utils to support passing ndarrays in a REST request
* modified unit test for Ludwig server to support passing ndarray images
* updated all other feature `add_feature_data()` methods for adding `skip_save_processed_input` parameter.

w4nderlust commented on ludwig/features/image_feature.py:
```diff
@@ -294,32 +314,45 @@ def add_feature_data(
                     num_processes > 1 or num_images > 1):
-                all_file_paths = [get_abs_path(src_path, file_path)
-                                  for file_path in input_df[feature[NAME]]]
+                all_img_stores = [get_abs_path(src_path, img_store)
```
maybe we can rename `img_store` to `img_entry` (`img_entries` and `all_image_entries`) for consistency

jimthompson5802 replied:
made change.

w4nderlust commented on ludwig/features/image_feature.py:
```diff
@@ -250,12 +258,24 @@ def add_feature_data(
             raise ValueError('There are no images in the dataset provided.')
 
-        first_path = next(iter(input_df[feature[COLUMN]]))
+        first_img_source = next(iter(input_df[feature[COLUMN]]))
```
We may call this `first_img_entry` for consistency

jimthompson5802 replied:
made change

w4nderlust commented on ludwig/features/image_feature.py:
```diff
@@ -152,8 +156,8 @@ def _read_image_and_resize(
-            preprocessing_parameters,
-            first_image_path
+            preprocessing_parameters: dict,
+            first_img_source: Union[str, 'numpy.array']
```
We may call this first_img_entry for consistency

jimthompson5802 replied:
done

w4nderlust commented on ludwig/features/image_feature.py:
```diff
@@ -58,16 +59,17 @@ def get_feature_meta(column, preprocessing_parameters, backend):
-            num_channels,
-            resize_method,
-            user_specified_num_channels
+            img_source: Union[str, 'numpy.array'],
```
img_source -> img_entry

jimthompson5802 replied:
done

jimthompson5802 commented:
Summary of changes:
* Deprecated the the ludwig custom string format for ndarrays and removed the related helper functions from `ludwig.utils.data_utils`.
* Complete re-write of the functions that serialize/deserialize the data used in the REST API.  Removed the old functions.  Created three new helper functions for the write of serializing and deserializing the REST API data.  Housed these new helper functions in a new module `ludwig.utils.server_utils.py`.  
* Addressed review comments on variable naming, i.e., `img_store` -> `img_entry`.

Work remaining:
* Have to add additional test for when `in_memory` and file/ndarray settings are different in training vs when predicting.
* Code clean-up on now obsolete code that is currently commented out.

jimthompson5802 commented:
Converted back to draft mode.  Clarification of requirements is forcing a rethink of current approach.
